contracts/src/spaces/facets/gated/EntitlementGatedBaseV2.sol

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+// interfaces
+import {IEntitlementGatedBaseV2} from "./IEntitlementGatedV2.sol";
+import {IRuleEntitlementV2} from "contracts/src/spaces/entitlements/rule/IRuleEntitlementV2.sol";
+import {IEntitlementChecker} from "contracts/src/base/registry/facets/checker/IEntitlementChecker.sol";
+import {IImplementationRegistry} from "contracts/src/factory/facets/registry/IImplementationRegistry.sol";
+
+// libraries
+import {EntitlementGatedStorageV2} from "./EntitlementGatedStorageV2.sol";
+import {MembershipStorage} from "contracts/src/spaces/facets/membership/MembershipStorage.sol";
+
+abstract contract EntitlementGatedBaseV2 is IEntitlementGatedBaseV2 {
+  function _setEntitlementChecker(
+    IEntitlementChecker entitlementChecker
+  ) internal {
+    EntitlementGatedStorageV2.layout().entitlementChecker = entitlementChecker;
+  }
+
+  function _requestEntitlementCheck(
+    bytes32 transactionId,
+    IRuleEntitlementV2 entitlement,
+    uint256 roleId
+  ) internal {
+    EntitlementGatedStorageV2.Layout storage ds = EntitlementGatedStorageV2
+      .layout();
+
+    Transaction storage transaction = ds.transactions[transactionId];
+
+    if (transaction.hasBenSet == true) {
+      for (uint256 i = 0; i < transaction.roleIds.length; i++) {
+        if (transaction.roleIds[i] == roleId) {
+          revert EntitlementGated_TransactionCheckAlreadyRegistered();
+        }
+      }
+    }
+
+    // if the entitlement checker has not been set, set it
+    if (address(ds.entitlementChecker) == address(0)) {
+      _setFallbackEntitlementChecker();
+    }
+
+    address[] memory selectedNodes = ds.entitlementChecker.getRandomNodes(5);
+
+    if (!transaction.hasBenSet) {
+      transaction.hasBenSet = true;
+      transaction.entitlement = entitlement;
+      transaction.clientAddress = msg.sender;
+    }
+
+    transaction.roleIds.push(roleId);
+
+    for (uint256 i = 0; i < selectedNodes.length; i++) {
+      transaction.nodeVotesArray[roleId].push(
```

Comment:
use EnumerableSet not arrays

Reply:
This is a straight copy of previous implementation, do you think we could defer to a follow-up PR?